### PR TITLE
Bash documentation

### PR DESCRIPTION
**Pull request by kevgo**
@charlierudolph @allewun @ricmatsui 

I think it makes sense to document how to use Bash. This is a start.

### CODE REVIEW

**Comment by charlierudolph** [comment]
I don't think we need to host this type of information ourselves. I'd prefer we just link to a good source on bash programming and document the patterns we choose to follow that aren't caught already by the linter.

**Comment by kevgo** [comment]
There are several reasons why I think adding this is useful:
* We have the issue that people are hesitant to join the project because they don't know Bash, and think it is too boring or complicated. Giving them a quick sheet that shows how simple and easy it actually is encourages participation and PRs.
* I'm not aware of another place that would provide this information in an nice and concise format. Most Bash documentation looks horrible and very old. If you know about one, please provide a link.
* Seeing all the things that are possible or not possible, and how exactly to do them helps us as developers. Looping correctly over arrays requires some serious syntax. I don't want to have to comb through the various Bash docs and StackOverflow pages to find the version that works here.
* This can serve as some sort of style guide, so that we do things only one way, therefore reducing entropy of the code base. 

In the end, having this doesn't cost us anything, and it provides a number of technical, psychological, and social benefits, so why not have it?

File: documentation/bash_cheatsheet.md [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+* [If clauses](#if-clauses)
+* [Functions](#functions)
+* [Arrays](#arrays)
+* [String Lists](#string-lists)
+* [Converting](#converting)
+
+
+## Shebang
+
+Use the cross-platform version of shebang:
```
Comment (allewun) [inline documentation/bash_cheatsheet.md]: Linkify: `[shebang](https://en.wikipedia.org/wiki/Shebang_%28Unix%29)`

File: documentation/bash_cheatsheet.md [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+
+  ```bash
+  if [ ! condition ]; then
+    # runs if the condition is not true
+  ```
+* checking booleans
+
+  ```bash
+  if [ $success = true ]; then
+    # :)
```
Comment (allewun) [inline documentation/bash_cheatsheet.md]: remove smile :frowning:   (for brevity and consistency with other bullet points)

File: documentation/bash_cheatsheet.md [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+  ```bash
+  function add_user {
+    local name=$1
+    echo "adding user $name"
+  }
+  ```
+
+  Notice how the first parameter is given a name
+  by assigning it
+  to the local variable `bar`.
```
Comment (allewun) [inline documentation/bash_cheatsheet.md]: to the local variable `$name`

File: documentation/bash_cheatsheet.md [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+* looping over the elements
+
+  ```bash
+  for person in "${people[@]}"; do
+    echo "$person"
+  done
+
+  # looping with index
+  for index in ${!people[*]}; do
+    printf "%4d: %s\n" $index ${people[$index]}
```
Comment (allewun) [inline documentation/bash_cheatsheet.md]: Let's remove the `4` to keep it simple

File: documentation/bash_cheatsheet.md [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+  ```bash
+  function containsElement {
+    local e
+    for e in "${@:2}"; do [[ "$e" == "$1" ]] && echo true && return; done
+    echo false
+  }
+  ```
+* returning arrays from methods
+
+  Methods can only return exit codes,
```
Comment (allewun) [inline documentation/bash_cheatsheet.md]: this paragraph seems misplaced

File: documentation/bash_cheatsheet.md [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+  no data structures.
+  Use global variables to pass around arrays.
+
+
+## String Lists
+
+Lists are the universal exchange format between unix programs.
```
Comment (allewun) [inline documentation/bash_cheatsheet.md]: UNIX (caps)

File: documentation/bash_cheatsheet.md [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+Bash is a surprisingly powerful, versatile, and elaborate scripting system
+where every command-line application becomes a keyword in your programming language.
+
+With the right amount of structure,
+Bash programs can be as easily written and read
+as programs in many other languages.
+
+Bash is the
+[9th most popular language on GitHub](http://githut.info),
+and actively used in most open-source repositories.
```
Comment (allewun) [inline documentation/bash_cheatsheet.md]: ... and [is] actively used...

File: documentation/bash_cheatsheet.md (hunk 5 above)
Comment (kevgo) [inline documentation/bash_cheatsheet.md]: Why? This paragraph explains why array cannot be returned from methods.

**Comment by kevgo** [comment]
Good comments, all of them addressed!

File: documentation/bash_cheatsheet.md (hunk 5 above)
Comment (allewun) [inline documentation/bash_cheatsheet.md]: I misread this, my bad. I was in auto-pilot mode and expected a code block.

File: documentation/bash_cheatsheet.md [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+  else
+    # code
+  fi
+  ```
+
+* negation
+
+  ```bash
+  if [ ! condition ]; then
+    # runs if the condition is not true
```
Comment (allewun) [inline documentation/bash_cheatsheet.md]: Can we remove this line too? I like your short one-line format that you used below.

File: documentation/bash_cheatsheet.md [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+  ```
+* checking booleans
+
+  ```bash
+  if [ $success = true ]; then
+    # runs if $success is true
```
Comment (allewun) [inline documentation/bash_cheatsheet.md]: here too

**Comment by allewun** [comment]
Should there be a link to this from another page? Looks good otherwise

**Comment by charlierudolph** [comment]
@kevgo please address @allewun comments then this is good to merge.

File: documentation/bash_cheatsheet.md (hunk 8 above)
Comment (kevgo) [inline documentation/bash_cheatsheet.md]: removed

File: documentation/bash_cheatsheet.md (hunk 9 above)
Comment (kevgo) [inline documentation/bash_cheatsheet.md]: removed

**Comment by kevgo** [comment]
Okay, I have also moved this into the /documentation/development folder, and linked to it from CONTRIBUTING.md.

**Pull request closed without merging**